File: Leetcode/tree/124_BinaryTreeMaximumPathSum/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        int[] res = {Integer.MIN_VALUE};
        dfs(root, res);
        return res[0];
    }
    private int dfs(TreeNode root, int[] res){
        if(root==null)
            return 0;
        int left = dfs(root.left, res), right = dfs(root.right, res);
        int returnVal = Math.max(root.val, Math.max(root.val+left, root.val+right));
        res[0] = Math.max(res[0], Math.max(returnVal, left+root.val+right));
        return returnVal;
    }
}

File: Leetcode/tree/124_BinaryTreeMaximumPathSum/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        res = [-sys.maxsize]
        self.dfs(root, res)
        return res[0]

    def dfs(self, root, res):
        if not root:
            return 0
        left = self.dfs(root.left, res)
        right = self.dfs(root.right, res)
        returnVal = max(root.val, max(root.val+left, root.val+right))
        res[0] = max(res[0], max(returnVal, left+root.val+right))
        return returnVal
